PSA API for PBKDF2-HMAC

mpg opened the pull request:
Resolves #4003.

This is a proposed addition to the PSA Cryptography API specification, which we intend to implement in Mbed TLS ahead of its official adoption by PSA.

Supporting documents:

* [password-hashing-abstract-api.md](https://github.com/ARMmbed/mbedtls/files/6343115/password-hashing-abstract-api.md)
* [password-hashing-psa-api.md](https://github.com/ARMmbed/mbedtls/files/6343116/password-hashing-psa-api.md)

Note: the supporting documents explore a much wider range of algorithms, in order to ensure that the API defined for PBKDF2 doesn't prevent us from extending to more key stretching / password hashing schemes in the future, but the focus of this PR is PBKDF2 only.

mpg commented:
@gilles-peskine-arm @athoelke @MarcusJGStreets @stevew817 any feedback from you would be highly appreciated!

In particular, I wasn't very familiar with existing conventions for allocating numeric values, I've try to do my best to follow existing conventions when I notice them, but may have failed in some cases. I've left questions and comments in the commit messages when I wasn't sure or considered several options.

athoelke reviewed (changes requested):
A number of inline comments in the changes, and a large overall comment: I am not convinced that we need a fully parallel set of key types, key usage flags **and** kdf input steps. Although I can see some of the underlying reasoning: breaking it down:

* We already reuse `PSA_KEY_USAGE_VERIFY_MESSAGE` for both verifying a MAC or asymmetric signature of a message; and have `PSA_KEY_USAGE_VERIFY_HASH` for verifying the signature of a message digest. Neither really fit the usage for verification of a hashed password. Although VERIFY_HASH _sounds_ like it fits the hashed password use case - the role the key plays in the two use cases is completely different so recycling the permission might be confusing. So I agree that a new permission for 'password' keys to be verified makes sense.
* The addition of a permission for a stored key to be used in password verification is also valid. There is no precedent here, as all existing verification APIs take an input buffer parameter. Perhaps password authentication is special? - or maybe we will add APIs to verify hashes, macs, and signatures against stored data in future? Do we need to scope this permission to just passwords, or can we just use `PSA_KEY_USAGE_VERIFIER` anticipating such changes? - or see my comment below about using a key type instead.

* Is the addition of `PSA_KEY_TYPE_PASSWORD` as a distinct type to provide a clear distinction within stored keys between arbitrary [non-secret?] data (RAW_DATA), low-entropy secrets (PASSWORD) and high-entropy secrets (DERIVE)? - or is there another reason.
* How valuable is the provision of a distinct key type for pepper, rather than reusing RAW_DATA? Or is pepper a shared secret, which warrants a distinction from RAW_DATA?
* Is RAW_DATA an appropriate key type to use for hashed passwords (definitely secret) retained in the key store? Or should we actually have another key type `PSA_KEY_TYPE_VERIFIER` for this instead of using a usage flag to signify what this key is?

* Do we need a new input step for input of a password to a password hashing KDF? - given that `PSA_KEY_DERIVATION_INPUT_SECRET` has exactly the right semantics (the password is meant to be secret). Elsewhere, we haven't created new input step names for a KDF algorithm just because the name better matches the normative algorithm definition. On the other hand, step names are just mappings to integer values, interpreted by each algorithm.

gilles-peskine-arm commented on include/psa/crypto_values.h:
```diff
@@ -2083,10 +2165,38 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ * psa_key_derivation_verify_output_bytes() or
+ * psa_key_derivation_verify_output_key() at the end of the operation.
+ */
+#define PSA_KEY_USAGE_PASSWORD_HASH_AND_VERITY  ((psa_key_usage_t)0x00008000)
```
(Using this thread to reply to @athoelke)

> The addition of a permission for a stored key to be used in password verification is also valid. There is no precedent here, as all existing verification APIs take an input buffer parameter. Perhaps password authentication is special?

Yes. For the primary purpose of a password hash, both the password and the hash are confidential. If the hash is leaked, this allows an adversary to conduct offline attacks to brute-force the password. If the adversary can only make queries of the form “is this password correct?”, then the attack has to be online, and so can be rate-limited, logged, and amount-limited. This is why a password hash is stored in a confidential database.

This confidential database may or may not be the PSA key store: depending on the deployments and on where storage is available, I expect that sometimes password hashes will be in the PSA key store (e.g. for a single-user IoT device that allows the user to authenticate with a password) and sometimes they will be in a user database outside the PSA world and PSA will only be used as a crypto implementation (e.g. for a multi-user server using PSA crypto because it has exclusive access to a hardware accelerator).

gilles-peskine-arm commented on include/psa/crypto_values.h:
```diff
@@ -414,6 +414,20 @@
  */
 #define PSA_KEY_TYPE_DERIVE                         ((psa_key_type_t)0x1200)
 
+/** A low-entropy secret for password hashing or key derivation.
```
(Using this thread to reply to @athoelke)

> Is the addition of PSA_KEY_TYPE_PASSWORD as a distinct type to provide a clear distinction within stored keys between arbitrary [non-secret?] data (RAW_DATA), low-entropy secrets (PASSWORD) and high-entropy secrets (DERIVE)?

Yes. Using a password with a cryptographic mechanism that requires high-entropy input is a very common mistake, and a goal of the PSA crypto API is to make such mistakes hard. It's the same reason the key agreement interface directs users to using a KDF on the raw KA output, but for passwords the consequences of misuse are much worse and so the API is stricter.

gilles-peskine-arm commented on include/psa/crypto_values.h:
```diff
@@ -414,6 +414,20 @@
+ */
+#define PSA_KEY_TYPE_PASSWORD                       ((psa_key_type_t)0x1300)
+
+/** A secret value that can be mixed in when doing password hashing.
```
(Using this thread to reply to @athoelke)

> How valuable is the provision of a distinct key type for pepper, rather than reusing RAW_DATA? Or is pepper a shared secret, which warrants a distinction from RAW_DATA?

A pepper is a share secret.

`RAW_DATA` is intended for data that has no direct cryptographic relevance or structure, at least as far as the PSA API is concerned. (A common use is to store a certificate alongside a private key. This has cryptographic relevance, but the crypto API doesn't use the certificate object, which is why it hasn't gotten its own type, though it might in a future version.) Even for a salt, using `RAW_DATA` is pushing it: a salt does have cryptographic “structure” in that it's meant to be unique, and it definitely has cryptographic relevance as the input to a cryptographic algorithm.

gilles-peskine-arm commented on include/psa/crypto_values.h:
```diff
@@ -2107,6 +2217,20 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ * may not be used to derive keys: the operation will only allow
+ * psa_key_derivation_output_bytes(), not psa_key_derivation_output_key().
+ */
+#define PSA_KEY_DERIVATION_INPUT_PASSWORD   ((psa_key_derivation_step_t)0x0102)
```
(Using this thread to reply to @athoelke)

> Do we need a new input step for input of a password to a password hashing KDF?

I think that for password hashing we could get away without it, but the distinction is also useful for some PAKE.

gilles-peskine-arm reviewed (changes requested):
Overall this is in good shape. There are a few places that could use minor tweaks to the design or editorial improvements.

stevew817 commented:
Before making this PR close #4003, can we either add an algorithm identifier for `PSA_ALG_PBKDF2_AES_CMAC_PRF_128` (or another ID construct which would support doing that through this interface), or add another task to cover that?

Thread is dependant on PBKDF2 (ref #806), so this specific function being exercisable / accelerated through the PSA API would be very much appreciated.

mpg commented on include/psa/crypto.h:
```diff
@@ -3290,6 +3290,48 @@ psa_status_t psa_key_derivation_input_bytes(
+ * Refer to the documentation of each key derivation or key agreement
+ * algorithm for information.
+ *
+ * This function is used for inputs which are small non-negative integers.
```
I wasn't entirely happy with "small" either but had trouble coming up with anything better. Thanks for suggesting "fixed-size", I'll use that.

mpg commented on include/psa/crypto_values.h:
```diff
@@ -414,6 +414,20 @@
+ * The key policy determines which key derivation algorithm the key
+ * can be used for.
+ */
+#define PSA_KEY_TYPE_PASSWORD                       ((psa_key_type_t)0x1300)
```
Sure. As the commit message said, I had no idea what the existing conventions were, so I just started with an initial guess and relied on reviewers who knew better to correct my mistakes. For my own education, is this scheme documented somewhere (and also the requirement that key types have even parity)?

athoelke replied:
Yes, the encoding scheme is planned for inclusion as an appendix in the next update of the PSA Crypto spec.

gilles-peskine-arm replied:
Only in private discussions in https://github.com/ARMmbed/psa-crypto-api I'm afraid.

mpg commented on include/psa/crypto.h:
```diff
@@ -3601,6 +3643,100 @@ psa_status_t psa_key_derivation_output_key(
+ *         The output was read successfully, but if differs from the expected
+ *         output.
+ * \retval #PSA_ERROR_NOT_PERMITTED
+ *         The key passed as the expected value does not allow this usage.
```
I have no strong opinion, but I will note that when you store a password hash for the purpose of password verification, it only makes sense to use it with a single algorithm: the one that was used to produce it in the first place. So don't think it would be an unreasonable constraint on applications to require the permitted algorithm to also be set for this key.

gilles-peskine-arm replied:
Yes, the permitted algorithm of the stored hash should match the hash used to hash the putative password. Otherwise a stored hash could be used to authenticate based on an algorithm that is so broken it allows preimage attacks.

The key _is_ an input to the stretch-hash-and-compare process. So using the algorithm field would not be novel.

mpg commented on include/psa/crypto_values.h:
```diff
@@ -1667,6 +1699,56 @@
+ * known as key stretching or password hashing schemes. These are also the
+ * algorithms that accepts inputs of type #PSA_KEY_DERIVATION_INPUT_PASSWORD.
+ */
+#define PSA_ALG_KEY_DERIVATION_STRETCHING_FLAG  ((psa_algorithm_t)0x00008000)
```
I'm sorry, this is a bit beyond me. Could you explain again like I'm not familiar with the context?

gilles-peskine-arm replied:
For better or for worse, an algorithm can encode both a key agreement and a key derivation, for example `PSA_ALG_KEY_AGREEMENT(PSA_ALG_ECDH, PSA_ALG_HKDF(PSA_ALG_SHA_256))`. To use such an algorithm:
* `psa_key_derivation_setup` with the algorithm
* `psa_key_derivation_key_agreement` to perform ECDH
* Further `psa_key_derivation_xxx` to perform HKDF where the secret input is the shared secret from the ECDH part.

A key stretching has its own way of inputting the secret which is not compatible with a key agreement. In the encoding of key derivation and key agreement algorithms, bits 0..15 encode the KDF and bits 16..23 encode the key agreement. A key stretching doesn't need to be compatible with a key agreement, so the encoding of key stretching algorithms can use bits 16..23. Key derivation algorithms that are meant to be usable for the output of a key agreement can't use bits 16..23 in their encoding.

Therefore I propose to use a bit from 16..23 to indicate “this is a key stretching KDF”. There's a separate bit in 24..31 that indicates “this is a key derivation without a key agreement”, so there's no ambiguity.

athoelke replied:
It would be good to ensure that if an application incorrectly attempted to combine a key agreement with a stretching KDF, that the result is not a known/defined composite algorithm. At present, the encoding scheme has zero in bits 23 and 22 for a key agreement, and encodes the key agreement type in bits 21:16.

Thus using a value that sets one or both of bits 23 and 22 to one should have the desired property. I think my preference is to designate bit 23 (as in @gilles-peskine-arm 's suggestion) as the key-stretching-KDF indicator. (All of the currently specified KDFs expect high-entropy secrets and so match this addition to the encoding scheme).

mpg commented on include/psa/crypto_values.h:
```diff
@@ -1667,6 +1699,56 @@
+ * known as key stretching or password hashing schemes. These are also the
+ * algorithms that accepts inputs of type #PSA_KEY_DERIVATION_INPUT_PASSWORD.
+ */
+#define PSA_ALG_KEY_DERIVATION_STRETCHING_FLAG  ((psa_algorithm_t)0x00008000)
```
Question: I think the "specification-defined values" you're referring to is https://armmbed.github.io/mbed-crypto/html/appendix/specdef_values.html but where can I find this "encoding appendix"?

gilles-peskine-arm replied:
https://github.com/ARMmbed/psa-crypto-api/pull/468

mpg commented on include/psa/crypto_values.h:
```diff
@@ -1667,6 +1699,56 @@
+ */
+#define PSA_ALG_IS_PBKDF2_HMAC(alg)                                    \
+    (((alg) & ~PSA_ALG_HASH_MASK) == PSA_ALG_PBKDF2_HMAC_BASE)
+#define PSA_ALG_PBKDF2_HMAC_GET_HASH(hkdf_alg)                         \
```
Looks like we don't have it in Mbed TLS yet, and instead we have a bunch of `PSA_ALG_xxx_GET_HASH()` so I followed the pattern I was seeing. Looking at appendix C, I see this changed between beta3 and 1.0.0.

I'm thinking perhaps I should leave this until we add `PSA_ALG_GET_HASH()` then? @gilles-peskine-arm wdyt?

mpg commented on include/psa/crypto.h:
```diff
@@ -3601,6 +3643,100 @@ psa_status_t psa_key_derivation_output_key(
+ *         It is implementation-dependent whether a failure to initialize
+ *         results in this error code.
+ */
+psa_status_t psa_key_derivation_verify_output_bytes(
```
I'm not sure. On one hand, `_output_` hints that some output will be consumed from the key derivation operation. OTOH, shorter names are nice too. @athoelke wdyt?

athoelke replied:
I agree with @gilles-peskine-arm - `psa_key_derivation_verify_bytes()` feels better.

mpg commented on include/psa/crypto.h:
```diff
@@ -3601,6 +3643,100 @@ psa_status_t psa_key_derivation_output_key(
+ *
+ * This function calculates output bytes from a key derivation algorithm and
+ * compares those bytes to an expected value, provided as key of type
+ * #PSA_KEY_TYPE_RAW_DATA.
```
I'm not sure what distinction you're doing between `PASSWORD_HASH` and `DERIVATION_OUTPUT`, could you clarify?

gilles-peskine-arm replied:
By “password hash”, I mean the output of a key stretching KDF, and by “key derivation output”, the output of a non-stretching KDF. I make a distinction because it's feasible to recover the password from its hash if the password isn't very strong, whereas it's infeasible to recover the high-entropy secret input to a KDF from its output.

gilles-peskine-arm replied:
> [Do we need a new key type (or even two) other than RAW_DATA for the key input of verify_output_key() and how should we call it?](https://github.com/ARMmbed/mbedtls/pull/4377#issuecomment-830021213)

We at least need a type for password hashes, which is suitable as the output of `psa_key_derivation_output_key` on a key stretching function and as the input of `psa_key_derivation_verify_key`. Internally, I this key might even store the parameters used to generate the hash (salt, cost) as well as the raw output. I suggest the name `PSA_KEY_TYPE_PASSWORD_HASH`.

By analogy, it would make sense to have a dedicated type `PSA_KEY_TYPE_DERIVATION_OUTPUT` for the non-stretching case. However, unlike the strecthing case, I don't think it's a common scenario.

Generally, avoid `PSA_KEY_TYPE_RAW_DATA` for anything that has a cryptographic “meaning”: anything that has a specific structure (e.g. some asymmetric keys), randomness or uniqueness requirements (keys, nonces, …), etc. `PSA_KEY_TYPE_RAW_DATA` is ok for a label that has no generic expectation of secrecy or uniqueness.

mpg replied:
I'm not entirely sure we need that distinction here. Imagine a hash `h` is produced using a password hashing scheme `A`, so we know `h = A(p)` for some `p` in a relatively small set `S` (which we assume the attack can enumerate easily). Now assume an attacker can somehow trick us into using a different algorithm B for verification, that was not designed to handle passwords and is much faster for the attacker to evaluate. There's no reason there would exist a set `S'` that the attacker could easily enumerate and that would necessarily contain a `p'` such that `h = B(p')`.

So, really I think even if we wanted to support verification in the non-stretching case, I'm not sure a different key type would be necessary.

Anyway, for now I've introduced a single key type called `PASSWORD_HASH` as this is the thing that's used in real life, and out of precaution I've documented it as being usable only with password hashing algorithms. If in the future we find out people are doing that with non-stretching KDFs too, at this point we can always decide to either (1) keep only the existing type and lift the documented restriction - the name would just reflect the most common use case, or (2) keep the restriction on the existing type and add a second one.

So, I think the way it's done in the current version of this PR is safe and does not block further extension to the non-stretching case, should it become desirable.

mpg commented on include/psa/crypto.h:
```diff
@@ -3601,6 +3643,100 @@ psa_status_t psa_key_derivation_output_key(
+ * \param[in,out] operation The key derivation operation object to read from.
+ * \param[in] expected      A key of type #PSA_KEY_TYPE_RAW_DATA containing
+ *                          the expected output. Its policy must include the
+ *                          #PSA_KEY_USAGE_PASSWORD_HASH_VERIFIER flag.
```
This is probably due to me lacking familiarity with the existing API: I didn't think there was anything special about creating a key of type `RAW_DATA`. Oh, looking again at the documentation of `output_key()` I can see it has an explicit list of possible key types and `RAW_DATA` is not one of them, so indeed I should have added it, good point.

I'll postpone that after we've decided whether we need a different key type (or even two) here.

athoelke commented on include/psa/crypto.h:
```diff
@@ -3601,6 +3643,100 @@ psa_status_t psa_key_derivation_output_key(
+ *         It is implementation-dependent whether a failure to initialize
+ *         results in this error code.
+ */
+psa_status_t psa_key_derivation_verify_output_bytes(
```
but as suggested already: perhaps `psa_key_derivation_verify_key()`?

athoelke commented on include/psa/crypto_values.h:
```diff
@@ -786,6 +800,24 @@
 #define PSA_ALG_IS_KEY_DERIVATION(alg)                                  \
     (((alg) & PSA_ALG_CATEGORY_MASK) == PSA_ALG_CATEGORY_KEY_DERIVATION)
 
+/** Whether the specified algorithm is a key stretching / password hashing
```
Ok. In that case, my preference would be to use "or" rather than "/".

mpg commented:
@athoelke @gilles-peskine-arm Thanks for the thorough reviews! I've just pushed a version that addresses most of your feedback. Unless I've missed something, the only points I have not addressed yet are:

1. Do we need a new key type (or even two) other than `RAW_DATA` for the key input of `verify_output_key()` and how should we call it? Gilles, I'm not sure I understand why you're suggesting two types here, or even that correctly understand what you're suggesting.
2. Policy checks for using a key with `verify_output_key()`: just the flag discussed above, or must also specify the algorithm? I'm inclined towards also specifying the algorithm, but I might be missing some context about existing conventions here.
3. Usage names: can we use `USAGE_VERIFY_DERIVATION` instead of the two `USAGE_PASSWORD_xxx` flags I was suggesting? I think that would be better indeed.
4. Function names: `verify_output_xxx()` or simply `verify_xxx()`? I'm on the fence.

Please feel free to weigh in (or clarify your position) on those questions. I'll make the corresponding changes on Monday.

gilles-peskine-arm commented:
1. https://github.com/ARMmbed/mbedtls/pull/4377#discussion_r624163801
2. https://github.com/ARMmbed/mbedtls/pull/4377/files#r624169107
3. https://github.com/ARMmbed/mbedtls/pull/4377#discussion_r619894476 → yes
4. https://github.com/ARMmbed/mbedtls/pull/4377#discussion_r623793432, https://github.com/ARMmbed/mbedtls/pull/4377#discussion_r616777490 → yes

mpg commented:
@athoelke @gilles-peskine-arm I believe I have now addressed all of your feedback so far, and now the PR passes CI too, so it's ready for you to review again.

athoelke commented on include/psa/crypto_values.h:
```diff
@@ -2107,6 +2217,20 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ * The secret can also be a direct input (passed to
+ * key_derivation_input_bytes()). In this case, the derivation operation
+ * may not be used to derive keys: the operation will only allow
+ * psa_key_derivation_output_bytes(), not psa_key_derivation_output_key().
```
Will the operation allow `psa_key_derivation_verify_bytes()`? - this is not mentioned in the text.

athoelke commented on include/psa/crypto_values.h:
```diff
@@ -2083,10 +2224,34 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ * - for a key of type #PSA_KEY_TYPE_PASSWORD_HASH, as the \c key argument of
+ *   psa_key_derivation_verify_key();
+ * - for a key of type #PSA_KEY_TYPE_PASSWORD (or #PSA_KEY_TYPE_DERIVE), as
+ *   the input to psa_key_derivation_input_key() at the step
```
Documentation for `psa_key_derivation_input_key()` needs to be updated to permit keys with this usage flag (currently only `PSA_KEY_USAGE_DERIVE` is permitted).

athoelke reviewed (changes requested):
A couple of missing connections (one normative, one informative) are needed in the documentation.

* https://github.com/ARMmbed/mbedtls/pull/4377#discussion_r625704515
* https://github.com/ARMmbed/mbedtls/pull/4377#discussion_r625708006

stevew817 commented on include/psa/crypto_values.h:
```diff
@@ -1804,6 +1933,18 @@
+ *
+ * \return #PSA_ALG_NONE if alg is not a composite algorithm that uses a hash.
+ */
+#define PSA_ALG_GET_HASH(alg) \
```
If there's now a generic 'get hash component if there is one' macro, is it then expected to replace all of the other `_GET_HASH` instances?
If so, they should be moved to deprecated status.
If not, then this one should probably be named `PSA_ALG_PBKDF2_HMAC_GET_HASH` to be in-line with the other `GET_HASH`'s

gilles-peskine-arm replied:
[`PSA_ALG_GET_HASH`](https://armmbed.github.io/mbed-crypto/html/api/ops/algorithms.html?highlight=psa_alg_get_hash#c.PSA_ALG_GET_HASH) is in the 1.0.0 spec. Adding it is part of https://github.com/ARMmbed/mbedtls/issues/3278. We've decided to add it now so that we wouldn't need to introduce a `PSA_ALG_PBKDF2_HMAC_GET_HASH` only to deprecate it soon afterwards.

stevew817 replied:
OK, thanks for the explanation.

stevew817 reviewed (changes requested):
LGTM, besides the `PSA_ALG_GET_HASH` macro which I don't believe is or should be a part of the PSA spec?

mpg commented:
Thanks @stevew817 and @athoelke for your reviews! While addressing Andrew's comment on `USAGE_VERIFY_DERIVATION` in the documentation of `input_key` I noticed that things were not very well documented as to what's allowed now that `input_key()` allows several usages and the verification is done in one of the "finishing" functions (`output_xxx` or `verify_xxx`), so I've added commit attempting to clarify that.

I think the implementation can use the following strategy:
- Keep the existing `can_output_key` bit (possibly renamed `can_finish_with_key`), handled the same way: initially set to `0`, changed to `1` by `psa_key_derivation_input_key()` when it's used at step `INPUT_SECRET` or `INPUT_PASSWORD`. This bit controls usage of `psa_key_derivation_{output,verify}_key()`.
- Add a bit `can_ouput` which is initially 1, and is changed to `0` by `psa_key_derivation_input_key()` when it sees a key that doesn't allow `PSA_KEY_USAGE_DERIVE`. This bit controls usage of `psa_key_derivation_output_xxx()`.
- Add a bit `can_verify` which is initially 1, and is changed to `0` by `psa_key_derivation_input_key()` when it sees a key that doesn't allow `PSA_KEY_USAGE_VERIFY_DERIVATION`. This bit controls usage of `psa_key_derivation_verify_xxx()`.

This is perhaps a bit more strict than necessary (for example, `verify_xxx()` could be allowed every time `output_xxx()` is), but at least it's somewhat orthogonal and consistent, and I think it encourages good practice (being consistent with respect to direct inputs vs keys, using the minimal policy that supports what you want to do).

Does that sound reasonable?

gilles-peskine-arm commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -305,9 +305,12 @@ This family requires the following type and entry points:
 * `"key_derivation_input_key"` (opaque drivers only)
 * `"key_derivation_output_bytes"`: called by `psa_key_derivation_output_bytes()`; also by `psa_key_derivation_output_key()` for transparent drivers.
 * `"key_derivation_output_key"`: called by `psa_key_derivation_output_key()` for transparent drivers when deriving an asymmetric key pair, and also for opaque drivers.
+* `"key_derivation_verify_output_bytes"` (opaque drivers only).
+* `"key_derivation_verify_output_key"` (opaque drivers only).
```
```suggestion
* `"key_derivation_verify_bytes"` (opaque drivers only).
* `"key_derivation_verify_key"` (opaque drivers only).
```

gilles-peskine-arm commented on include/psa/crypto_values.h:
```diff
@@ -408,12 +408,56 @@
+ * This key type is suitable for passwords and passphrases which are typically
+ * intended to be memorizable by humans, and have a low entropy relative to
+ * their size. It can be used for randomly generated or derived keys with
+ * maximum or near-maximum entropy, but PSA_KEY_TYPE_DERIVE is more suitable
```
```suggestion
 * maximum or near-maximum entropy, but #PSA_KEY_TYPE_DERIVE is more suitable
```

gilles-peskine-arm commented on include/psa/crypto_values.h:
```diff
@@ -2083,10 +2165,38 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ * psa_key_derivation_input_key() at the step
+ * #PSA_KEY_DERIVATION_INPUT_SECRET of #PSA_KEY_DERIVATION_INPUT_PASSWORD
+ * depending on the algorithm, and allows the use of
+ * psa_key_derivation_output_bytes() or psa_key_derivation_output_key()
```
This is not addressed. Strictly speaking the language here doesn't claim that the list of functions is exhaustive, but it's a bit misleading not to list the verify functions.

mpg replied:
Well, actually in the current version of this PR, the verify functions need the `VERIFY_DERIVATION` bit, as documented in the error section of these functions and in the note on `input_key()`. I don't have a strong opinion on it remaining that way, but let me share the reasoning that led to the current state so that we can assess it and decide whether we change it or not.

If the `DERIVE` bit allows everything, people might be tempted to use it everywhere for simplicity and not think whether `VERIFY_DERIVATION` would suffice for what they're doing. Also (or perhaps it's just another way of saying the same thing?), it seemed a bit more orthogonal/consistent to me for each set of functions (output, derive) to be independently controlled by its own usage bit, rather than one usage bit being a superset of the other.

Again, I'm not married to this line of reasoning, just spelling out my thought process so that we can discuss it.

gilles-peskine-arm replied:
Thinking about it again, I agree with your reasoning. Looking at derivation alone, it doesn't really make sense to allow output() without allowing verify(). However, it's just up to the application not to do that. For other operations, we decided to keep things simple and have one permission flag map to one operation, and for example you can't verify a MAC with `USAGE_VERIFY` alone even though you can calculate the MAC and compare it. So key derivation should follow the same principle.

There is however one reason to make derivation different, which is backward compatibility. Suppose I've created a key derivation key with an implementation of PSA Crypto 1.0.0, and it has `USAGE_DERIVE` because that's all there is. It happens that my application calls `output_bytes` only to then compare these bytes with a reference value. So now I want to call `verify_bytes` instead, just to simplify my code, but I can't.

I'm not convinced that this is a compelling reason, so let's go for the simple way now (no change).

gilles-peskine-arm commented on include/psa/crypto_values.h:
```diff
@@ -2102,11 +2267,32 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
+ * psa_key_derivation_output_bytes() or
+ * psa_key_derivation_verify_bytes(), not
+ * psa_key_derivation_output_key() or
+ * psa_key_derivation_verify_key().
```
Thinking about it, I don't think it makes sense to prohibit `psa_key_derivation_verify_key` here. I expect it to be a common use case for password verification: the user has just entered a password, which the caller has in memory, so the input comes via `input_bytes` or requires the unnecessary overhead of a temporary key; the password hash is a non-exportable key object, so the output has to be consumed by `verify_key`.

This workflow is less common for non-password inputs, but I think `INPUT_SECRET` and `INPUT_PASSWORD` should have the same rules in this respect, so this change should apply to `INPUT_SECRET` as well.

mpg replied:
I see your point. Initially I thought there was a parallel with account creation (if the caller has the password in main memory, they also need the unnecessary overhead of a temporary key for the password in order to be able to use `output_key()`: but now I realise not so much: `output_key()` can easily be replaced with `output_bytes()` + `import_key()`, while the same doesn't hold for `verify_key()` (assuming the key is not exportable).

gilles-peskine-arm replied:
The restriction on `output_key` is for application hygiene, not directly for security: if you're using some derived material as a key, and the material was derived from data that came from outside the crypto service, then the derived key itself comes from outside the crypto service. There's a desire (although I'm not sure how important this is) to allow implementations to keep track of which keys were always inside and which were potentially exposed, and this restriction very slightly simplifies `output_key`.

This hygiene restriction is on creating a key, so it only applies to `output_key`, not `verify_key`.

mpg replied:
Thanks, the reasoning behind the restriction on `ouput_key` was never entirely clear to me, which is probably one of the reasons I incorrectly transposed it to `verify_key`. It's much clearer now.

athoelke replied:
The approach is not 'obviously sensible', as it leaves asymmetries in the process for enrolling a new password and verifying it later. Supposing that the input password is held in application memory, then enrolment looks like one of the following sequences:

1. Setup KDF operation for PBKDF
2. Input cost parameters with `input_number()`
3. Input salt with `input_bytes()`
4. Input password with `input_bytes()`
5. Erase password buffer
6. Extract password hash with `output_bytes()`
7. Store password hash with `import_key()` (type = PASSWORD_HASH, usage = VERIFY_DERIVATION, alg = PBKDF)
8. Erase hash buffer

Or by using a temporary key for the password:

1. Create temporary key with password (type = PASSWORD, usage = VERIFY_DERIVATION, alg = PBKDF)
2. Erase password buffer
3. Erase password buffer
4. Setup KDF operation for PBKDF
5. Input cost parameters with `input_number()`
6. Input salt with `input_bytes()`
7. Input password with `input_key()`
8. Store password hash with `output_key()` (type = PASSWORD_HASH, usage = VERIFY_DERIVATION, alg = PBKDF)
9. Destroy temp password key

In contrast, verification doesn't have the extra juggling step:

1. Setup KDF operation for PBKDF
2. Input cost parameters with `input_number()`
3. Input salt with `input_bytes()`
4. Input password with `input_bytes()`
5. Erase password buffer
6. Verify password hash with `verify_key()`

I think the flow for password verification has less application hygiene risks than those which are related to deriving keys - for the flows above I do not see why denying the simplification of directly storing the password hash is better?

Could the restriction on `output_key()` be triggered by the input function used for `INPUT_SECRET`, but not by `INPUT_PASSWORD` or `INPUT_SALT/PEPPER` etc? - which might deal with another of the open issues in this discussion?

Technically, we could infer what the `output_key()` will be used for based on its type or its usage policy, which could permit password-hash-storing usage but not cryptographic-key-derivation usage. This feels like it might be one level of complexity too far (for comprehension by either the users or implementers)?

mpg replied:
I think what you're suggesting makes sense in principle but I agree that the main question is that of complexity. As a general rule, I'm inclined towards making things simple to document even if it means the API a bit more restrictive that necessary, at least for a first version. For one thing, it's easier to lift restrictions in the future than to add them, for backwards compatibility reasons. Also as the API starts being used and we get feedback from users, we'll be in a better position to judge whether the additional flexibility is worth the additional complexity.

So, I'm inclined to keep the restriction on `ouput_key()` unchanged for now, and only lift the one on `verify_key()` as suggested by Gilles.

gilles-peskine-arm reviewed (changes requested):
Looks almost good to me, just a few minor points remain.

mpg commented on include/psa/crypto_values.h:
```diff
@@ -1779,6 +1779,17 @@
+ *
+ * PBKDF2 is defined by PKCS#5, republished as RFC 8018 (section 5.2).
+ * This macro specifies the PBKDF2 algorithm constructed using the
+ * AES-CMAC-PRF-128 PRF specified by RFC 4615.
```
I'm sorry, perhaps I'm reading RFC 6803 too quickly but I don't think it's about using using a Camellia-based construct as the PRF for PBKDF2, but rather about using PBKDF2-HMAC-SHA1 to derive Camellia keys. Actually, they even need to tweak the salt to achieve domain separation from AES keys, which they wouldn't need to do if they were changing the PBKDF2 PRF.

Before adding this identifier, I had similar reservations so I spent some time looking for uses of other PRFs in PBKDF2 in publicly-available standards, open source software and other publicly documented APIs, and couldn't find any other than the ones in this PR, which is the reason to target specifically AES-128.

(Of course that doesn't prove there aren't any, I could have missed them and some might even appear in the future - I would think nowadays people would turn to newer schemes, but the Thread spec would tend to prove me wrong.) But overall, I think the probabily that we end up introducing other algorithm identifiers in the future seems low, and even then as you note, we'd probably end up adding a couple specific names rather than a generic macro anyway.

So I'm inclined to  keep this identifier here. But if after considering my points you still feel it's best to defer, I'm OK with that too.

gilles-peskine-arm replied:
Ah, yes, my bad: RFC 6803 uses a Camellia-CMAC-based non-stretching KDF to derive key material, but uses PBKDF2-HMAC-SHA1 to stretch.

So while I'm still a little uncomfortable with encoding one specific block cipher in an algorithm (because block ciphers are somewhat region-specific), I'm ok with the pragmatic choice of including this one because it's the only one that has a known deployed protocool using it.

mpg commented on include/psa/crypto.h:
```diff
@@ -3359,12 +3359,30 @@ psa_status_t psa_key_derivation_input_integer(
+ * \note Once all inputs steps are completed, the operations will allow:
+ * - psa_key_derivation_output_bytes() if each input was either a direct input
+ *   or  a key with #PSA_KEY_USAGE_DERIVE set;
+ * - psa_key_derivation_output_key() if each input was either a direct input
+ *   or  a key with #PSA_KEY_USAGE_DERIVE set and input for step
+ *   #PSA_KEY_DERIVATION_INPUT_SECRET or #PSA_KEY_DERIVATION_INPUT_PASSWORD
+ *   was from a key slot;
```
I think the first clause is more generic as it covers other inputs that might come from a key slot too - typically the pepper.

Note that currently the documentation of `input_key()` says the key "must have an appropriate type for step and must allow the usage `PSA_KEY_USAGE_DERIVE`." To me it reads like the usage requirement applies to all input steps where you choose to read the input from a key, not just `INPUT_SECRET`. So I tried to keep the same logic here.

I agree that the current wording is not very clear, so I'll see if I can find a better one - suggestions welcome too.

mpg commented:
@gilles-peskine-arm I think we're close to converging, but not fully there yet when it comes to policies and when each function should be allowed. I think [this message](https://github.com/ARMmbed/mbedtls/pull/4377#issuecomment-832531145) and [this commit](https://github.com/ARMmbed/mbedtls/pull/4377/commits/ece9087b9375b841b724d5f5551fc8d0347c4254) sum up my intentions as of yesterday.

I think there are a few points still under consideration:

1. We could lift the requirement that the secret/password was input from a key to allow use of `verify_key()`. (You think we should, I don't disagree, even if that was not the choice I made yesterday.)
2. We need to decide if the verify functions require `USAGE_VERIFY_DERIVATION` or `USAGE_DERIVE` also enables them. 
3. We need to make sure we agree on usage requirements when multiple inputs come from keys: if (part of) the salt is passed using `input_key()` from a key that only allows `USAGE_VERIFY_DERIVATION` and the password is passed from a key that has `USAGE_DERIVE`, should it be allowed to call the `output` functions? I think not. But I only realised yesterday that this was a question at all, so my reasoning might not be very mature on this point.

I'd like to reach/clarify an agreement on those points before I push the next version of the PR. @athoelke feel free to weigh in as well.

gilles-peskine-arm commented on include/psa/crypto.h:
```diff
@@ -3314,12 +3358,31 @@ psa_status_t psa_key_derivation_input_bytes(
  * \param key                     Identifier of the key. It must have an
  *                                appropriate type for step and must allow the
- *                                usage #PSA_KEY_USAGE_DERIVE.
+ *                                usage #PSA_KEY_USAGE_DERIVE or
```
I wonder if the specification was overly restrictive here. The _secret_ (`PSA_KEY_DERIVATION_INPUT_SECRET`) input to a key derivation should be a suitable key, i.e. it should have the type `PSA_KEY_TYPE_DERIVE` for `input_key` or a key agreement type for `key_agreement`, and should have the usage `PSA_KEY_USAGE_DERIVE`. But it can be useful to take a key object as a salt/label/… input, too, regardless of whether it allows `USAGE_DERIVE`.

However not all keys should be usable that way. For example, passwords must not be usable in this way, since it would amount to allowing them to be hashed by a non-stretching key derivation!

mpg replied:
I think the key type avoids that. Currently `input_key()` requires the type of the key to be compatible with the step, and the only step that accepts a key of type `PASSWORD` is `INPUT_PASSWORD`. Other steps accept keys or type `RAW_DATA`, or `PEPPER` for the salt step.

gilles-peskine-arm replied:
Right. But a salt still has to have `USAGE_DERIVE`, which I'm not convinced is a useful requirement.

mpg replied:
I agree, I'm not convinced it's useful either, but I don't think it's hurtful either (that's what my previous comment was about), I didn't think about it enough to be fully 100% convinced it's not useful, it's pre-existing, and lifting requirements is a backwards-compatible change that can be done later, so I'd rather keep it as is in this PR and leave it for later. Would you object to that?

gilles-peskine-arm commented:
@mpg

1. [I maintain my change request.](https://github.com/ARMmbed/mbedtls/pull/4377#discussion_r626808792) I think it's a compelling use case.
2. [Let's go your way.](https://github.com/ARMmbed/mbedtls/pull/4377#discussion_r627471761)
3. [I don't know](https://github.com/ARMmbed/mbedtls/pull/4377/files#r627484836). I'm ok with making it go either way for now and keeping that an open question.

athoelke commented:
How big is the output of a password-hashing/key-stretching algorithm, when used for password verification? - i.e. what size should be used for `output_bytes()`, and what size key should be specified for `output_key()`? - is there a standard output size for a given algorithm, or can the user decide how much hash security they want?

mpg commented:
Some algorithms (Balloon, bcrypt) have a fixed-size output. For the rest, the output length for password hashing is not always specified or even recommended by the standard defining the algorithm. NIST has a set of recommendations in [SP 800-63b](https://pages.nist.gov/800-63-3/sp800-63b.html) which mentions PBKDF2, but I could only find recommendations about salt size and number of iterations, not about about output size.

I *think* for PBKDF2-HMAC it makes sense to use the output size of the underlying hash as the output size, but that's really just an educated guess. For PBKDF2-AES-CMAC I would similarly guess using the block size of AES (that is, 128 bits) makes sense but that even more of a guess.

I there some standard or official recommendation that I missed here?

mpg commented:
Update: I found one library that does what I said regarding PBKDF2-HMAC: https://passlib.readthedocs.io/en/stable/lib/passlib.hash.pbkdf2_digest.html section "Format and algorithms":
> Each scheme uses the digest size of its specific hash algorithm (digest) as the size of the raw derived key.

Another weak signal in that direction is that in RFC 6070, most test vectors for PBKDF2-HMAC-SHA1 use an output length of 20 (the output size of SHA-1).

That's still not very conclusive.

mpg commented:
I've pushed a new version addressing most of the comments so far. Regarding the previous list of open points:

1. Lift restriction on usage of `verify_key()`: I've gone with Gilles's suggestion. I consider that one close unless someone objects.
2. Make `USAGE_DERIVE` a superset of `USAGE_VERIFY_DERIVATION` or keep them separate: I've kept them separate. I consider that one close unless someone objects.
3. Keep or lift the restriction that keys used at any step (not just the main secret) must allow the proper usage: I've kept the restriction. (Reasoning similar as [that other point](https://github.com/ARMmbed/mbedtls/pull/4377#discussion_r628089531).) I consider that one close unless someone objects.

I think new points have emerged:

4. Should we lift the restriction on `output_key()` under some condition to be determined (for example, the output key is of type `PASSWORD_HASH` with usage `VERIFY_DERIVATION`)? I'm inclined towards "[not now, perhaps later](https://github.com/ARMmbed/mbedtls/pull/4377#discussion_r628089531)".
5. Should we include a recommendation on output length for the purposes of password verification, and if so, what should we recommend? I'm not sure, I'll probably ask on Slack too.
6. (I'm adding this now, I though about it while summarizing 4 above.) When `output_key()` writes its output to a key of type `PASSWORD_HASH`, should it verify that the algorithm in the key policy matches the one in the operation?

Wdyt?

gilles-peskine-arm commented:
Regarding the size: PBKDF2 generates output in blocks whose size is that of the underlying hash. The expensive stretching calculation is done separately for each block. So for example generating 33 bytes of PBKDF2-HMAC-SHA-256 is twice as expensive as generating 32 bytes. So you generally would want to avoid generating more than one block with PBKDF2. I think all the more modern algorithm have a stretch-then-expand structure where the stretching cost is independent of the output length, so generating longer outputs only has a small additional cost independent of the cost factor.

This is something we'd want to document, but I don't think it has any impact on the API design. Unless we want some kind of `PSA_PASSWORD_HASH_RECOMMENDED_LENGTH(…)`?

gilles-peskine-arm commented:
> When output_key() writes its output to a key of type PASSWORD_HASH, should it verify that the algorithm in the key policy matches the one in the operation?

Yes, good point.

gilles-peskine-arm reviewed (dismissed):
Looks good to me other than [the issue Andrew raised](https://github.com/ARMmbed/mbedtls/pull/4377#discussion_r630105965).

gilles-peskine-arm commented on include/psa/crypto_values.h:
```diff
@@ -2117,7 +2297,8 @@ static inline int mbedtls_svc_key_id_is_null( mbedtls_svc_key_id_t key )
  * This should be a direct input.
- * It can also be a key of type #PSA_KEY_TYPE_RAW_DATA.
+ * It can also be a key of type #PSA_KEY_TYPE_RAW_DATA or
+ * #PSA_KEY_TYPE_PEPPER.
```
Argon2 officially takes both a salt and a pepper. For other algorithms, when a pepper is used, it's a part of the salt, with no particular standard to put it first or last.

athoelke replied:
> Our first priority is PBKDF2. If pepper needs tuning, we can deal with it later.

So - should we remove `PSA_KEY_TYPE_PEPPER` until we need it for another algorithm?

athoelke replied:
Now I just remembered the discussion that we already had about salt and pepper: typical usage for algorithms like PBKDF2 is to concatenate the per-application secret pepper and per-id non-secret salt and provide them as the "salt" to PBKDF2.

Hence using the INPUT_SALT with a new key type, mostly to help the developer pay attention to what the data in the "key" is. The proposed API design just reflects the evolved use (salt+pepper) of existing password hashing functions (which only knew about salt).

The API might still look cleaner if we provided as `INPUT_PEPPER` step (for use with `PSA_KEY_TYPE_PEPPER` keys), and algorithms such as PBKDF2 documented that any `INPUT_PEPPER` inputs are concatenated in sequence with any `INPUT_SALT` inputs?

gilles-peskine-arm replied:
No, we should keep pepper: we know it's useful, including for PBKDF2. And yes, for most algorithms including PBKDF2, pepper and salt are the same input. Since a pepper is secret, it should have its own type.

mpg replied:
Ok, so it seems we agree about keeping `KEY_TYPE_PEPPER`.

I think there's a separate question of whether we should use `INPUT_PEPPER` for PBKDF2. (I think it's quite clear that when we add Argon2, it will have `INPUT_PEPPER`, so below I'm just focusing on the question of introducing it for things like PBKDF2.) I think we need to have a special case somewhere, and introducing `INPUT_PEPPER` or not determines where this special case lives.

1. If we introduce `INPUT_PEPPER`, the nice thing is that `INPUT_SALT` keeps using `KEY_TYPE_RAW_DATA` like all the other non-secret inputs. The special case is that (for all algs except Argon2), `INPUT_PEPPER` is not a separate input but is concatenated with other `INPUT_SALT` inputs; and we still have to document that `INPUT_SALT` can be repeated (in case some scheme uses a salt structure like `per_entry_salt || pepper || context`). Also, the list of input steps accepted by PBKDF2 will look different to the descriptions of its input in the standard.
2. If we don't introduce `INPUT_PEPPER`, the nice thing is that the description of the input steps matches the standard. The special case is that `INPUT_SALT` must be documented as accepting both secret inputs (with `KEY_TYPE_PEPPER`) and non secret inputs (with `KEY_TYPE_RAW_DATA` or more likely `input_bytes()`). (Plus that `INPUT_SALT` can be repeated, but that's actually common with the other case.)

I think option 2, which is the current state of the PR, better represents the standard (input steps match it) + usage that evolved on top of it (key type tells the _nature_ of the data passed at this step). Though I'm open to hearing arguments for option 1 if you're not convinced!

gilles-peskine-arm replied:
We discussed in https://github.com/ARMmbed/psa-crypto-api/pull/489 and came to the conclusion that it was best to stick to `INPUT_SALT` for algorithms that don't have a dedicated pepper input, because the pepper would be concatenated into the salt. I don't see any new arguments in this discussion, so let's stick to what we have.

mpg commented on include/psa/crypto.h:
```diff
@@ -3601,6 +3672,129 @@ psa_status_t psa_key_derivation_output_key(
+ *
+ * This function calculates output bytes from a key derivation algorithm and
+ * compares those bytes to an expected value, provided as key of type
+ * #PSA_KEY_TYPE_RAW_DATA, in constant time.
```
Aw sorry, I missed that one when changing that key type. Thanks for catching it! I just searched the full diff for other occurrences of `RAW_DATA` and didn't find any other issue, so I think you caught the last one.

stevew817 reviewed (approved):
I took a look at the changes since my last approval, and have no more comments.

gilles-peskine-arm commented:
win32-mingw failed in pr-merge due to an infrastructure glitch, but it passed in pr-head so we can take the CI as a pass.

silabs-ArchanaM commented:
@mpg @gilles-peskine-arm 
There could be hardware acceleration support where the whole PBKDF2_hmac/cmac derivation would be supported by hardware.
In cases such as that, how do we justify splitting the routines into input/options and output.
Would it not sit well to also have a single API to do this, as opposed to having to split across multiple APIs for input options and output.

The current design works well if you're inherently offloading the underlying algorithm alone right ?

mpg commented:
Hi @silabs-ArchanaM ! I'm not sure I understand this part of your question : " if you're inherently offloading the underlying algorithm alone right" - do you mean if we're offloading just the HMAC or CMAC part as opposed to PBKDF2 itself?

I don't think the design makes that kind of assumption. What I was assuming is that for each input/option step, the driver just copies the info to some internal structure, and then when output is called, the driver passes everything to the hardware in order to get some output bytes back. (Note that even if you only offload HMAC or CMAC, or even have a purely software implementation, I think that's also the most natural implementation strategy.)

The reason we have the functions split as they are in the _user_ API is that we want the API to be work with multiple KDFs / password hashing schemes (PHS). Since they all take different numbers and types of inputs and parameters, having a single C function would not be convenient. Hence the current design, which I think is good for the _user_ API.

The reason we have the functions split in the _driver_ API is that by default it tends to mirror the user API. Now here I agree that it could make sense to have the core collect the inputs & parameters, and then pass that in one go to the driver function. It would mean you'd need to have a different driver for each KDF / PHS. Also, there's the question of getting the output in multiple chunks: we'd need a "operation" type visible by the core, for each supported KDF / PHS driver. That could probably be done, but since the existing KDF driver API was mirroring the user API, I chose to keep the same design for the PHS driver API.

I hope this answers your question, if I misunderstood something please let me know.
